src/AbstractMenuElement.php

```diff
@@ -0,0 +1,213 @@
+    }
+
+    /*
+     * @since [*next-version*]
+     */
+    use ParentAwareTrait;
+
+    /*
+     * @since [*next-version*]
+     */
+    use CapabilityAwareTrait;
+
+    /*
+     * @since [*next-version*]
+     */
+    use IconAwareTrait;
+
+    /**
+     * Parameter-less constructor.
+     *
+     * Call this in the real constructor.
+     *
+     * @since [*next-version*]
+     */
+    protected function _construct()
+    {
+        if (empty($this->children)) {
+            $this->children = [];
+        }
+    }
+
+    /**
+     * Adds a child menu at a given position.
+     *
+     * @since [*next-version*]
+     *
+     * @param mixed    $child    The menu element to add as a child.
+     * @param int|null $position An integer, with larger values signifying a lower position, or null to append.
+     *
+     * @return $this
+     */
+    protected function _addChild($child, $position = null)
+    {
+        $this->_validateChild($child);
+
+        if ($position === null) {
+            $this->_appendChild($child);
+
+            return $this;
+        }
+
+        // Normalize position - detects collisions and generates new keys
+        $position = $this->_normalizePosition($position, $child->getKey(), $child->getLabel());
+
+        // Add child at determined position
+        $this->children["$position"] = $child;
+
+        // Keep the children sorted
+        ksort($this->children);
+
+        return $this;
+    }
+
+    /**
+     * Checks whether this menu element has children.
+     *
+     * @since [*next-version*]
+     *
+     * @return bool True if this instance has children; false otherwise.
+     */
+    protected function _hasChildren()
+    {
+        $children = $this->_getChildren();
+        $count    = 0;
+
+        if (is_array($children) || $children instanceof Countable) {
+            $count = count($children);
+        }
+
+        if ($children instanceof \Traversable) {
+            $count = iterator_count($children);
+        }
+
+        return $count > 0;
+    }
+
+    /**
+     * Checks whether this instance has a parent.
+     *
+     * @since [*next-version*]
+     *
+     * @return bool True if this instance has a parent; false otherwise.
+     */
+    protected function _hasParent()
+    {
+        return $this->_getParent() !== null;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    protected function _validateChild($child)
+    {
+        if (!$this->_isMenuElement($child)) {
+            throw $this->_createValidationFailedException('Child is not a valid menu element.', 0, null, $child);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    protected function _validateParent($parent)
+    {
+        if ($parent !== null && !$this->_isMenuElement($parent)) {
+            throw $this->_createValidationFailedException('Parent is not a valid menu element.', 0, null, $parent);
```

Comment:
Actually, menus are not meant to be parent-aware. I had forgotten to remove the trait.